File: python/socketd/socketd/transport/stream/impl/RequestStreamImpl.py
import asyncio
from typing import Callable

from socketd.exception.SocketDExecption import SocketDTimeoutException, SocketDException
from socketd.transport.core.Costants import Constants
from socketd.transport.core.Entity import Reply
from socketd.transport.core.Message import MessageInternal
from socketd.transport.stream.RequestStream import RequestStream
from socketd.transport.stream.impl.StreamBase import StreamBase
from socketd.utils.CompletableFuture import CompletableFuture
from socketd.utils.RunUtils import RunUtils


class RequestStreamImpl(StreamBase, RequestStream):
    def __init__(self, sid: str, timeout: int):
        super().__init__(sid, Constants.DEMANDS_SINGLE, timeout)
        self.__future: CompletableFuture = CompletableFuture()

    def is_done(self):
        return self.__future.done()

    def __await__(self):
        try:
            return self.__future.get(self.timeout() / 1000.0)
        except asyncio.TimeoutError as _e:
            raise SocketDTimeoutException(f"Request reply timeout>{self.timeout()}  sid={self.sid()}")
        except Exception as _e:
            raise SocketDException(f"Request failed, sid= sid={self.sid()} {str(_e)}")

    async def waiter(self) -> Reply:
        if self.__future.done():
            return self.__future.get_result()
        return await self.__await__()

    def on_reply(self, reply: MessageInternal):
        self.__future.accept(reply)

    def on_error(self, error: Exception):
        super().on_error(error)
        self.__future.cancel()

    def then_reply(self, onReply: Callable[[MessageInternal], None]) -> RequestStream:
        async def _then_reply_do(r: MessageInternal, e: Exception):
            if r:
                await RunUtils.waitTry(onReply(r))
        self.__future.then_async_callback(_then_reply_do)
        return self

    def then_error(self, onError: Callable[[Exception], None]) -> RequestStream:
        self._then_error_do(onError)
        return self

    def then_progress(self, onProgress: Callable[[bool, int, int], None]) -> RequestStream:
        self._then_progress_do(onProgress)
        return self

File: javascript/socketd/src/socketd/transport/stream/impl/RequestStreamImpl.ts
import {IoConsumer} from "../../core/Typealias";
import {Reply} from "../../core/Entity";
import {Constants} from "../../core/Constants";
import {MessageInternal} from "../../core/Message";
import {RequestStream} from "../Stream";
import {StreamBase} from "./StreamBase";

/**
 * 请求流
 *
 * @author noear
 * @since 2.0
 */
export class RequestStreamImpl extends StreamBase<RequestStream> implements RequestStream{
    _doOnReply:IoConsumer<Reply>;
    _isDone:boolean;
    constructor(sid:string,  timeout:number) {
        super(sid, Constants.DEMANDS_SIGNLE, timeout);
        this._isDone = false;
    }

    isDone(): boolean {
        return this._isDone;
    }

    onReply(reply: MessageInternal) {
        this._isDone = true;

        try {
            if(this._doOnReply){
                this._doOnReply(reply);
            }
        } catch (e) {
            this.onError(e);
        }
    }

    await(): Promise<Reply> {
        return new Promise<Reply>((resolve, reject) => {
            this.thenReply(reply => {
                resolve(reply);
            }).thenError((err) => {
                reject(err);
            })
        })
    }

    thenReply(onReply: IoConsumer<Reply>): RequestStream {
        this._doOnReply = onReply;
        return this;
    }
}